### ISSUE TITLE
Change enhanceApp context API

### ISSUE THREAD

**Issue by IlyaSemenov** [ream/ream#125]
TL;DR — I propose to change enhanceContext API for clarity.

Below I listed all current public API contexts:

### createApp

```
{
  req
  res
}?
```

### entry

```
{
  req
  res
}?
```

### enhanceApp

There's actually two arguments:

```
{
  rootOptions
  entry
  ssrContext?: { req, res, entry }
  event
  getInitialDataContext
  addMiddleware
},
{ req, res, entry }?
```

### createApp response

```
{
  app
  entry
  router
  getInitialDataContextFns
  event
  dataStore
  middlewares
}
```

### getInitialData context, middleware context

```
{
  req?
  res?
  router
  route
  error
  redirect
}
```

---

I think before it's too late `enhanceApp` should be streamlined to to have 1 plain object argument:

```
{
  req?
  res?
  router
  rootOptions // should NOT have router, it will be injected at the last moment
  entry
  event
  addInitialData
  addMiddleware
}
```

I'm not sure about `addInitialData` naming, ideally it should be consistent with middleware, hence the proposal. The current naming of `getInitialDataContext` is very misleading, it's far from obvious that calling it doesn't *get* you anything.

**Comment by IlyaSemenov** [comment]
By the way: it just occurred to me that `getInitialDataContextFns` and `middlewares` are doing **exactly the same job**. They are always called each after another and never used separately.

On server:

```js
  for (const fn of getInitialDataContextFns) {
    fn(dataContext)
  }

  await runMiddlewares(middlewares, dataContext)
```

On client:

```js
    const ctx = getContext({ route: to, router, ...serverHelpers }) // getContext loops over getInitialDataContextFns 
    await runMiddlewares(middlewares, ctx)
```

and:

```js
      const context = getContext({
        router,
        route: to
      })

      await runMiddlewares(middlewares, context)
```

I propose to get rid of `getInitialDataContext` / `getInitialDataContextFns` and only keep `addMiddleware` / `middlewares`.

**Comment by egoist** [comment]
btw what about renaming `getInitialData` to simply `initialData` 🤔

**Comment by IlyaSemenov** [comment]
That's you baby — I can't really tell. Both sound good to me. `initialData` is perhaps a bit better because it's more consistent with `data` but then this renaming will break all existing projects 🤔 By the way, another option would be `asyncData` for easier transition from Nuxt.

Perhaps we may allow `getInitialData` for older code but have it emit a deprecation warning? We will need a universal getter instead of direct property access because of #101 anyway, which could handle that in one place.

**Comment by IlyaSemenov** [comment]
Closed in 5107f5a9d42df012d427a9af479f4c189d0a9e44 and ceeaa369291f2330bfa1ace17e705db2413d8611, published as v4.0.0.
